server/block/smelter.go

```diff
@@ -0,0 +1,176 @@
+func (s *smelter) Inventory() *inventory.Inventory {
+	return s.inventory
+}
+
+// AddViewer adds a viewer to the furnace, so that it is updated whenever the inventory of the furnace is changed.
+func (s *smelter) AddViewer(v ContainerViewer, _ *world.World, _ cube.Pos) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.viewers[v] = struct{}{}
+}
+
+// RemoveViewer removes a viewer from the furnace, so that slot updates in the inventory are no longer sent to
+// it.
+func (s *smelter) RemoveViewer(v ContainerViewer, _ *world.World, _ cube.Pos) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if len(s.viewers) == 0 {
+		// No viewers.
+		return
+	}
+	delete(s.viewers, v)
+}
+
+// tickSmelting ticks the smelter, ensuring the necessary items exist in the furnace, and then processing all inputted
+// items for the necessary duration.
+func (s *smelter) tickSmelting(requirement, decrement time.Duration, lit bool, supported func(item.SmeltInfo) bool) bool {
+	s.mu.Lock()
+
+	prevCookDuration := s.cookDuration
+	prevRemainingDuration := s.remainingDuration
+	prevMaxDuration := s.maxDuration
+
+	input, _ := s.inventory.Item(0)
+	fuel, _ := s.inventory.Item(1)
+	product, _ := s.inventory.Item(2)
+
+	var inputInfo item.SmeltInfo
+	if i, ok := input.Item().(item.Smelt); ok && supported(i.SmeltInfo()) {
+		inputInfo = i.SmeltInfo()
+	}
+
+	var fuelInfo item.FuelInfo
+	if f, ok := fuel.Item().(item.Fuel); ok {
+		fuelInfo = f.FuelInfo()
+		if fuelInfo.Residue.Empty() {
+			fuelInfo.Residue = fuel.Grow(-1)
+		}
+	}
+
+	canSmelt := input.Count() > 0 && (inputInfo.Product.Comparable(product)) && !inputInfo.Product.Empty() && product.Count() < product.MaxCount()
+	if s.remainingDuration <= 0 && canSmelt && fuelInfo.Duration > 0 && fuel.Count() > 0 {
+		s.remainingDuration, s.maxDuration, lit = fuelInfo.Duration, fuelInfo.Duration, true
+		defer s.inventory.SetItem(1, fuelInfo.Residue)
+	}
+
+	if s.remainingDuration > 0 {
+		s.remainingDuration -= time.Millisecond * 50
+		if canSmelt {
+			s.cookDuration += time.Millisecond * 50
+			if s.cookDuration >= requirement {
+				defer s.inventory.SetItem(2, item.NewStack(inputInfo.Product.Item(), product.Count()+inputInfo.Product.Count()))
+				defer s.inventory.SetItem(0, input.Grow(-1))
+
+				xp := inputInfo.Experience * float64(inputInfo.Product.Count())
+				earned := math.Floor(inputInfo.Experience)
+				if chance := xp - earned; chance > 0 && rand.Float64() < chance {
+					earned++
+				}
+
+				s.cookDuration -= requirement
+				s.experience += int(earned)
+			}
+		} else if s.remainingDuration == 0 {
+			s.maxDuration = 0
+		} else {
+			s.cookDuration = 0
+		}
+	} else {
+		s.maxDuration, s.remainingDuration, lit = 0, 0, false
+	}
+
+	if s.cookDuration > 0 && !lit {
+		s.cookDuration -= decrement
+	}
+
+	for v := range s.viewers {
+		v.ViewFurnaceUpdate(prevCookDuration, s.cookDuration, prevRemainingDuration, s.remainingDuration, prevMaxDuration, s.maxDuration)
+	}
+
+	s.mu.Unlock()
```

Comment:
Would move this all the way to the top under `s.mu.Lock()` and change it to `defer s.mu.Unlock()`.

Reply:
Wouldn't this cause issues with the `SetItem` calls being called before `mu` is unlocked?

Reply:
@Sandertv